### PR TITLE
cli: add internal tables to expose local info about the cluster

### LINKED ISSUE #20713: cli/user,zone,node: run `user`, `zone`, `node` commands through SQL
Most `cli` commands individually query certain endpoints and compile a result which they then pretty-print (see, for example, [status node](https://github.com/cockroachdb/cockroach/blob/d1550657cb767a952e25d617c684a74bb3ebad8a/pkg/cli/node.go#L129)`.

This is suboptimal because it implies that certain kind of information is only exposed through the `cli`, and that extra development effort has to be undertaken to make the same status available through SQL when that is requested. Operators are usually able to connect to a cluster using SQL anyway, and using the `cli` for the same task may be cumbersome.

My proposal is to move the implementation of the `user`, `zone`, and `node` commands to SQL wholesale (so that the `cli` commands are just small wrappers that issue a SQL query).

Some of their information is already available in SQL and the rest should be, and in particular should be presented in a comparable way. For example, the various possible outputs of `./cockroach node status` should be simple joins of more elementary tables in `crdb_internal`.

Some care has to be taken to make sure that we support returning partial information (https://github.com/cockroachdb/cockroach/issues/16489). That is, it'd be unfortunate if we joined `crdb_internal.pretty_much_always_available_info_from_gossip` with `crdb_internal.this_only_works_when_cluster_healthy`. But these joins are small, and with just a hint of tooling the `cli` can do these in-memory while leaving blanks. Or, we make these tables itself return partial information, though that raises questions about timeouts, etc.

### PR DESCRIPTION
**Pull request by dsaveliev**
Fixes #20713.

Release note: None

### CODE REVIEW

**Comment by tschottdorf** [comment]
Some straightforward commentary, but at a whole, this is almost ready to go! BTW, you seem to have a merge conflict in `explain`, you should see it once you `git fetch origin master && git rebase origin/master`.

---

Reviewed 5 of 5 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[pkg/sql/crdb_internal.go, line 1598 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21017#-L0yxsOP9pOr5gAT7tEa:-L0yxsOP9pOr5gAT7tEb:b-gnn41p) ([raw file](https://github.com/cockroachdb/cockroach/blob/c29251bdde6d5cde134eed503a837652fe5884e7/pkg/sql/crdb_internal.go#L1598)):*
> ```Go
> 		}
> 		g := p.session.execCfg.Gossip
> 		for key, info := range g.GetInfoStatusByPrefix(gossip.KeyNodeLivenessPrefix).Infos {
> ```

You need to sort the livenesses (and similarly the node statuses). How about this?

```go
// IterateInfos visits all infos matching the given prefix. The visitor will be passed a
// copy of the original. When the visitor returns an error, iteration stops and the error
// is returned from IterateInfos.
func (g *Gossip) IterateInfos(prefix string, visit func(k string, info Info)) error {
	g.mu.Lock()
	defer g.mu.Unlock()
	for k, v := range g.mu.is.Infos {
		if strings.HasPrefix(k, prefix+separator) {
            if err := visit(k, *(protoutil.Clone(v).(*Info))); err != nil {
                return err
            }
		}
	}
    return nil
}

var livenesses []storage.Liveness
if err := g.IterateInfos(gossip.KeyNodeLivenessPrefix, func(k string, i gossip.Info) {
    bytes, err := info.Value.GetBytes()
    if err != nil {
        return errors.Wrapf(err, "failed to extract bytes for key %q", k)
    }

    var liveness storage.Liveness
    if err := protoutil.Unmarshal(b, &l); err != nil {
        return errors.Wrapf(err, "failed to parse value for key %q", k)
    }
    append(livenesses, liveness)
    return nil
}); err != nil {
    return err
}

sort.Slice(livenesses, func(i, j int) bool {
    return livenesses[i].NodeID < livenesses[j].NodeID
})

// `livenesses` is ready to be put into the table rows`
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21017#-:-L0z2rbxgjpTxdxLnjUI:bd1d80l)*

**Comment by dsaveliev** [comment]
Review status: 2 of 5 files reviewed at latest revision, 1 unresolved discussion.

---

*[pkg/sql/crdb_internal.go, line 1598 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21017#-L0yxsOP9pOr5gAT7tEa:-L1-B2DLK-Ul7xan3PjV:b-roo3b) ([raw file](https://github.com/cockroachdb/cockroach/blob/c29251bdde6d5cde134eed503a837652fe5884e7/pkg/sql/crdb_internal.go#L1598)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

You need to sort the livenesses (and similarly the node statuses). How about this?

```go
// IterateInfos visits all infos matching the given prefix. The visitor will be passed a
// copy of the original. When the visitor returns an error, iteration stops and the error
// is returned from IterateInfos.
func (g *Gossip) IterateInfos(prefix string, visit func(k string, info Info)) error {
	g.mu.Lock()
	defer g.mu.Unlock()
	for k, v := range g.mu.is.Infos {
		if strings.HasPrefix(k, prefix+separator) {
            if err := visit(k, *(protoutil.Clone(v).(*Info))); err != nil {
                return err
            }
		}
	}
    return nil
}

var livenesses []storage.Liveness
if err := g.IterateInfos(gossip.KeyNodeLivenessPrefix, func(k string, i gossip.Info) {
    bytes, err := info.Value.GetBytes()
    if err != nil {
        return errors.Wrapf(err, "failed to extract bytes for key %q", k)
    }

    var liveness storage.Liveness
    if err := protoutil.Unmarshal(b, &l); err != nil {
        return errors.Wrapf(err, "failed to parse value for key %q", k)
    }
    append(livenesses, liveness)
    return nil
}); err != nil {
    return err
}

sort.Slice(livenesses, func(i, j int) bool {
    return livenesses[i].NodeID < livenesses[j].NodeID
})

// `livenesses` is ready to be put into the table rows`
```
</blockquote></details>

I agree with that. 
An iterator was implemented.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21017)*

**Comment by tschottdorf** [comment]
:lgtm: Thanks!

---

Reviewed 3 of 3 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21017#-:-L1-OJuJfDwVmZMYy_Py:b-g2a5fb)*

File: pkg/sql/crdb_internal.go [hunk 1]
```diff
@@ -1533,3 +1537,111 @@ CREATE TABLE crdb_internal.zones (
...
+		sort.Slice(descriptors, func(i, j int) bool {
+			return descriptors[i].NodeID < descriptors[j].NodeID
+		})
+
+		for _, d := range descriptors {
+			if err := addRow(
+				tree.NewDInt(tree.DInt(d.NodeID)),
+				tree.NewDString(d.Address.NetworkField),
+				tree.NewDString(d.Address.AddressField),
+				tree.NewDString(d.Attrs.String()),
```
Comment (tschottdorf) [inline pkg/sql/crdb_internal.go]: Actually, as a toy test for https://github.com/cockroachdb/cockroach/issues/20713#issuecomment-353689807, try turning `attrs` into a JSON field.

I think you'll need `tree.MakeDJSON(s)` where `s` is an `[]interface{}` with each entry being backed by a `string` (`s := "myattr", s = append(s, interface{}{s})`).

File: pkg/sql/crdb_internal.go [hunk 2]
```diff
@@ -1533,3 +1537,111 @@ CREATE TABLE crdb_internal.zones (
...
 }
+
+// crdbInternalGossipNodes exposes local information about the cluster nodes.
+var crdbInternalGossipNodes = virtualSchemaTable{
+	schema: `
+CREATE TABLE crdb_internal.gossip_nodes (
+  node_id  INT NOT NULL,
+  network  STRING NOT NULL,
+  address  STRING NOT NULL,
+  attrs    STRING NOT NULL,
```
Comment (tschottdorf) [inline pkg/sql/crdb_internal.go]: and this would be `JSON NOT NULL`.

**Comment by tschottdorf** [comment]
Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21017)*

**Comment by tschottdorf** [comment]
Thanks @dsaveliev!

**Pull request merged**